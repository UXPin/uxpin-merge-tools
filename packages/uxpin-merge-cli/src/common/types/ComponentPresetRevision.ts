import { ComponentPreset } from '../../steps/serialization/component/presets/ComponentPreset';

export interface ComponentPresetRevision extends ComponentPreset {
  componentRevisionId:string;
  presetId:string;
  sortIndex:number;
}
