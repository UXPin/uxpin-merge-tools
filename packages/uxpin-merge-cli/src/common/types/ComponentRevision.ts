import { ComponentDefinition } from '../../steps/serialization/component/ComponentDefinition';

export interface ComponentRevision extends ComponentDefinitionPersistedPart {
  componentId:string;
  revisionId:string;
}

export type ComponentDefinitionPersistedPart = Pick<ComponentDefinition, ComponentPersistedProps>;

type ComponentPersistedProps = 'name' | 'info' | 'properties';
