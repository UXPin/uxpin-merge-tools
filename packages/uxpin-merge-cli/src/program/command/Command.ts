export enum Command {
  DUMP = 'dump',
  EXPERIMENT = 'experiment',
  INIT = 'init',
  GENERATE_PRESETS = 'generate-presets',
  PUSH = 'push',
  DELETE_VERSION = 'delete-version',
  SERVER = 'server',
  SUMMARY = 'summary',
}

export const DEFAULT_COMMAND: Command = Command.EXPERIMENT;
