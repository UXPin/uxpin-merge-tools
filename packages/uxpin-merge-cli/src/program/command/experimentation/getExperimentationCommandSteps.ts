import { Store } from '../../../utils/store/Store';
import { ExperimentProgramArgs } from '../../args/ProgramArgs';
import { Step } from '../Step';
import { experimentationBuildLibraryStep } from './steps/experimentationBuildLibraryStep';
import { experimentationCreateEpidStep } from './steps/experimentationCreateEpidStep';
import { experimentationOpenBrowserStep } from './steps/experimentationOpenBrowserStep';
import { experimentationPrepareMetadataStep } from './steps/experimentationPrepareMetadataStep';
import { experimentationRunNgrok } from './steps/experimentationRunNgrok';
import { experimentationRunServerStep } from './steps/experimentationRunServerStep';

export interface ExperimentationState {
  ngrokUrl?: string;
}

const defaultState: ExperimentationState = {};

export function getExperimentationCommandSteps(args: ExperimentProgramArgs): Step[] {
  const store: Store<ExperimentationState> = new Store(defaultState);

  return [
    experimentationBuildLibraryStep(args),
    experimentationRunNgrok(args, store),
    experimentationCreateEpidStep(args),
    experimentationPrepareMetadataStep(args),
    experimentationRunServerStep(args, store),
    experimentationOpenBrowserStep(args, store),
  ];
}
