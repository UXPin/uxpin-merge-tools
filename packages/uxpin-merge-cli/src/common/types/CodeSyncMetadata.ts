import { ComponentPresetRevision } from './ComponentPresetRevision';
import { ComponentRevision } from './ComponentRevision';

export interface CodeSyncMetadata {
  components:AllComponentsCollection;
  presets:AllPresetsCollection;
  bundles:AllBundlesCollection;
}

export interface AllComponentsCollection {
  [componentId:string]:ComponentRevision;
}

export interface AllPresetsCollection {
  [presetId:string]:ComponentPresetRevision;
}

export interface AllBundlesCollection {
  [fetchRevisionId:string]:string;
}
