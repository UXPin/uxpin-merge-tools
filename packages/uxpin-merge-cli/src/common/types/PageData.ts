import { CodeSyncMetadata } from './CodeSyncMetadata';

export interface PageData {
  code_sync?:Partial<CodeSyncMetadata> | { presets:never[], components:never[] };
  component_version:any;
  components?:{ [componentId:number]:PageContent };
  components_ids?:string[];
  components_master_ids?:number[];
  components_nested_versions?:{ [componentId:number]:string };
  components_versions?:number[];
  components_versions_map?:number[];
  is_component?:string;
  last_update:string | number;
  page:PageContent;
}

interface PageContent {
  canvas:{
    props:{
      storedElements:string[];
      [name:string]:any;
    };
    type:'Canvas',
    v:string | number;
  };

  [elementId:string]:any;
}
