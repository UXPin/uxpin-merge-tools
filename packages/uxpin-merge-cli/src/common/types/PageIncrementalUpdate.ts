import { PageContent } from './PageData';

export type DeletedElements = { [elementId: string]: true } | never[];

export interface PageIncrementalUpdate {
  can_break_cohesion?: boolean;
  changed_elements: Partial<PageContent>;
  comet_thread_id?: string;
  components_used?: any;
  deleted_elements: DeletedElements;
  design_system_used?: any;
  id_page: number | string;
  id_project: number | string;
  id_save: string;
  is_component?: '0' | '1';
}
