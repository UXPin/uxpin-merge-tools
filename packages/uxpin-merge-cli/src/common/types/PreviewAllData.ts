import { Breakpoint } from '../../steps/experimentation/server/common/breakpoints/breakpoints';
import { PageData } from './PageData';

interface PageTreeItem {
  documentationExists: boolean;
  id_page: number;
  is_active: boolean;
  main_version: number;
  name: string;
  parent: number | null;
  sort_order: number;
  version_of: number | null;
  version_type: number;
}

interface PageSize {
  height: number;
  scrollx: boolean;
  scrolly: boolean;
  width: number;
}

interface PageMetadata {
  content: string;
  size: PageSize;
}

interface PreviewPageData {
  canvasData: PageData;
  metaData: PageMetadata;
}

interface PreviewPagesData {
  [id: number]: PreviewPageData;
}

export interface PreviewAllData {
  breakpoints: Breakpoint[];
  pageId: number;
  pages: PageTreeItem[];
  pagesData: PreviewPagesData;
  redirect: boolean;
}
