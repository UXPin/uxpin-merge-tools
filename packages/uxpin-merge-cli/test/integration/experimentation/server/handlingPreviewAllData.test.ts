import { OK } from 'http-status-codes';
import { omit } from 'lodash';
import { Response } from 'request';
import { PageContent } from '../../../../src/common/types/PageData';
import { setTimeoutBeforeAll } from '../../../utils/command/setTimeoutBeforeAll';
import { setupExperimentationServerTest } from '../../../utils/experimentation/setupExperimentationServerTest';
import { getExpectedIntroPageWithExampleElementsGuessingUniqueIdsFrom } from './page/set/fixtures/getExpectedIntroPageWithExampleElementsGuessingUniqueIdsFrom';

const CURRENT_TIMEOUT = 300000;
setTimeoutBeforeAll(CURRENT_TIMEOUT);

describe('Experimentation mode - handling preview all data', () => {
  const port = 8888;
  let response: Response;
  const { request } = setupExperimentationServerTest({
    port,
    projectPath: 'resources/designSystems/twoComponentsWithConfig',
    serverCmdArgs: ['--webpack-config "./webpack.config.js"'],
  });

  beforeAll(async () => {
    const origin = 'https://app.uxpin.com';
    response = await request('/preview/all', { resolveWithFullResponse: true, headers: { origin } });
  });

  it('should responds with OK status code', async () => {
    expect(response.statusCode).toEqual(OK);
  });

  it('should responds with correct CORS headers and no-cache', async () => {
    // given
    const expectedHeaders: any = {
      'access-control-allow-credentials': 'true',
      'access-control-allow-headers': 'Origin, X-Requested-With, Content-Type, Accept, Range',
      'access-control-allow-origin': 'https://app.uxpin.com',
      'cache-control': 'no-cache',
      'content-type': 'application/json',
    };

    // then
    expect(response.headers).toEqual(expect.objectContaining(expectedHeaders));
  });

  it('should respond with proper body', () => {
    const responseBody: any = JSON.parse(response.body);
    expect(omit(responseBody, 'pagesData.1.canvasData.page')).toMatchSnapshot();
  });

  it('should contain correct page object in the body', () => {
    // when
    const responseBody: any = JSON.parse(response.body);
    const responsePage: PageContent = responseBody.pagesData['1'].canvasData.page;

    // then
    const expectedPage: PageContent = getExpectedIntroPageWithExampleElementsGuessingUniqueIdsFrom(responsePage);
    expect(responsePage).toEqual(expectedPage);
  });
});
