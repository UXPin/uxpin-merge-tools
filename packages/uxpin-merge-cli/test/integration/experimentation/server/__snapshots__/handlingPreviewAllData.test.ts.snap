// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Experimentation mode - handling preview all data should respond with proper body 1`] = `
Object {
  "breakpoints": Array [
    Object {
      "height": 768,
      "name": "Wide 1440",
      "version": 0,
      "width": 1440,
    },
    Object {
      "height": 480,
      "name": "iPhone (320px)",
      "version": 1,
      "width": 320,
    },
    Object {
      "height": 320,
      "name": "Phone Landscape (480px)",
      "version": 2,
      "width": 480,
    },
    Object {
      "height": 768,
      "name": "Tablets (600px)",
      "version": 3,
      "width": 600,
    },
    Object {
      "height": 1024,
      "name": "iPads (768px)",
      "version": 4,
      "width": 768,
    },
    Object {
      "height": 768,
      "name": "Standard websites (992px)",
      "version": 5,
      "width": 992,
    },
    Object {
      "height": 768,
      "name": "iPad Landscape (1024px)",
      "version": 6,
      "width": 1024,
    },
    Object {
      "height": 768,
      "name": "Wide websites (1224px)",
      "version": 7,
      "width": 1224,
    },
  ],
  "pageId": 1,
  "pages": Array [
    Object {
      "documentationExists": false,
      "id_page": 1,
      "is_active": true,
      "main_version": 0,
      "name": "Experiment",
      "parent": null,
      "sort_order": 0,
      "version_of": null,
      "version_type": 0,
    },
  ],
  "pagesData": Object {
    "1": Object {
      "canvasData": Object {
        "code_sync": Object {
          "bundles": Object {
            "3ab57996-fdf2-41cd-b3c6-85ba98596081_33a58bbfb9e97c671048f796c842723f13599762": "https://sessionId.ngrok.io/code/library.js",
          },
          "components": Object {
            "ba14886c-2674-52a3-a147-7b88e725e4ee": Object {
              "componentId": "ba14886c-2674-52a3-a147-7b88e725e4ee",
              "info": Object {
                "dirPath": "src/components/Button",
                "implementation": Object {
                  "framework": "reactjs",
                  "lang": "typescript",
                  "path": "src/components/Button/Button.tsx",
                },
                "presets": Array [
                  Object {
                    "path": "src/components/Button/presets/0-default.jsx",
                  },
                ],
              },
              "name": "Button",
              "properties": Array [
                Object {
                  "description": "",
                  "isRequired": true,
                  "name": "children",
                  "type": Object {
                    "name": "node",
                    "structure": Object {},
                  },
                },
                Object {
                  "description": "",
                  "isRequired": false,
                  "name": "isPrimary",
                  "type": Object {
                    "name": "boolean",
                    "structure": Object {},
                  },
                },
              ],
              "revisionId": "3ab57996-fdf2-41cd-b3c6-85ba98596081_33a58bbfb9e97c671048f796c842723f13599762",
              "wrappers": Array [],
            },
            "f2dff102-3d25-5174-b733-12c4e58fdd5d": Object {
              "componentId": "f2dff102-3d25-5174-b733-12c4e58fdd5d",
              "info": Object {
                "dirPath": "src/components/Avatar",
                "implementation": Object {
                  "framework": "reactjs",
                  "lang": "typescript",
                  "path": "src/components/Avatar/Avatar.tsx",
                },
                "presets": Array [
                  Object {
                    "path": "src/components/Avatar/presets/0-default.jsx",
                  },
                ],
              },
              "name": "Avatar",
              "properties": Array [
                Object {
                  "description": "",
                  "isRequired": true,
                  "name": "size",
                  "type": Object {
                    "name": "string",
                    "structure": Object {},
                  },
                },
                Object {
                  "description": "",
                  "isRequired": true,
                  "name": "imageUrl",
                  "type": Object {
                    "name": "string",
                    "structure": Object {},
                  },
                },
              ],
              "revisionId": "3ab57996-fdf2-41cd-b3c6-85ba98596081_33a58bbfb9e97c671048f796c842723f13599762",
              "wrappers": Array [],
            },
          },
          "presets": Object {
            "02943eee-ab9a-57c6-adf8-d5c4978cd0b1": Object {
              "componentRevisionId": "3ab57996-fdf2-41cd-b3c6-85ba98596081_33a58bbfb9e97c671048f796c842723f13599762_f2dff102-3d25-5174-b733-12c4e58fdd5d",
              "elements": Object {
                "1": Object {
                  "name": "Avatar",
                  "props": Object {
                    "imageUrl": "https://picsum.photos/100/100?image=1027",
                    "size": "small",
                  },
                },
              },
              "name": "default",
              "presetId": "02943eee-ab9a-57c6-adf8-d5c4978cd0b1",
              "rootId": "1",
              "sortIndex": 0,
            },
            "364b2288-4144-5962-8ad2-7c5ebc0ab2ae": Object {
              "componentRevisionId": "3ab57996-fdf2-41cd-b3c6-85ba98596081_33a58bbfb9e97c671048f796c842723f13599762_ba14886c-2674-52a3-a147-7b88e725e4ee",
              "elements": Object {
                "1": Object {
                  "name": "Button",
                  "props": Object {
                    "children": "Click me",
                    "isPrimary": true,
                  },
                },
              },
              "name": "default",
              "presetId": "364b2288-4144-5962-8ad2-7c5ebc0ab2ae",
              "rootId": "1",
              "sortIndex": 0,
            },
          },
        },
        "component_version": null,
        "components_master_ids": Array [],
        "components_versions": Array [],
        "components_versions_map": Array [],
        "is_component": "0",
        "last_update": "0",
        "variation_data": Object {
          "height": "",
          "scrollx": false,
          "scrolly": false,
          "selectedSize": "1440x",
          "text": "Wide 1440",
          "version": 0,
          "width": 1440,
        },
      },
      "metaData": Object {
        "content": "/api/projects/3ab57996-fdf2-41cd-b3c6-85ba98596081_33a58bbfb9e97c671048f796c842723f13599762/pages/1/content",
        "size": Object {
          "height": 768,
          "scrollx": false,
          "scrolly": false,
          "width": 1440,
        },
      },
    },
  },
  "redirect": false,
}
`;
